### DISCUSSION TITLE
Testing the SMB-C authentication

### DISCUSSION THREAD

**Discussion by oliver-steinbrecher** [Digitaler-Impfnachweis/certification-apis#14] [General]
For PVS integrators the [SMC-B flow](https://github.com/Digitaler-Impfnachweis/certification-apis/blob/master/SMCB-Authentication.md) is the primary way for performing an authentication.  This page provides an overview for testing the flow in an isolated manner:

###  Testing with a RU connector
In order to test this we're running a dedicated test environment which is integrated with the Telematik Infratrustructure RU (Referenz Umgebung):

OIDC Authorization Endpoint: 
`https://id.ru.impfnachweis.info/auth/realms/bmg-ti-certify/protocol/openid-connect/auth`

| Query-Param | Value |  Details |
| --- | --- |  --- | 
| client_id  | user-access-ti | value is going to differ in production - same entry for all integrators |
| redirect_uri | connector://authenticated | same value in all stages | 
| response_type  | code |  
| scope | openid | 
| nonce | 'random-value' | 

### Testing with Gematik Kops connector simulator
This configuration support the usage of the gematik connector simulator. 

KoPS Version: > = 3.1

OIDC Authorization Endpoint: 
`https://id.certify.demo.ubirch.com/auth/realms/bmg-ti-certify/protocol/openid-connect/auth`

| Query-Param | Value |  Details |
| --- | --- |  --- | 
| client_id  | user-access-ti-simulator |  |
| redirect_uri | connector://authenticated | same value in all stages | 
| response_type  | code |  
| scope | openid | 
| nonce | 'random-value' | 

After performing the the authentication the following api endpoint can be used to issue DCCs:   https://api-pvs.certify.demo.ubirch.com

**Comment by nicolaslang1** [comment]
Hello @oliver-steinbrecher ,

the Gematik Kops connector simulator doesn't work with client_id = "smcb-auth-simulator".

We only receive a challenge with "user-access-ti" as client_id. But in the end when i pass the signed challenge and the certificate the api returns "not authorised". So there is a bug in the "https://id.certify.demo.ubirch.com/auth/realms/ubirch-certify/protocol/openid-connect/auth" Api. Could you please update the correct client_id.

Kind regards 

Nicolas

**Comment by ajobs** [comment]
Hi there,

I can confirm this (or at least I'm stuck at the same place):

Use of client_id=smcb-auth-simulator results in a "400 Bad request" in Step 1.
With client_id=user-access-ti we get a x-auth-challenge but a "401 Unauthorized" on submitting the signed challenge and the certificate (Step 4)

Any tips on that?

--Andreas

**Comment by norganim** [comment]
We can confirm, KoPS connector simulator smcb auth flow is not working with smcb-auth-simulator.

**Reply by WePo64** [reply]
We can confirm this Problem too.

**Reply by TH-Pega** [reply]
Same problem (error at step 4), even with the new value "user-access-ti-simulator". Maybe we need another SMC-B (current "Impfzentrum KoPS, Cardnumber 80276991110000007846, valid until 10/10/2023")?

**Comment by alexanderwilms** [comment]
@oliver-steinbrecher 
We use KOPS 3.1.11. Valid SMC-B  (ICCSN 80276883110000016161 Gutfall VSDM). 
endpoint: "https://id.certify.demo.ubirch.com/auth/realms/ubirch-certify/protocol/openid-connect/auth"
client_id: "user-access-ti-simulator"

Auth-Challenge and auth-location can be obtained from service. Example: 
- auth-challenge: 3dff2b0d-fd48-4252-9e81-fde901f72050
- auth-location: https://id.certify.demo.ubirch.com/auth/realms/ubirch-certify/login-actions/authenticate?execution=74acb014-15f9-4d47-8f28-c611bbe6c530&client_id=user-access-ti-simulator&tab_id=XSV0ZCffMM8

SMC-B certificate C-AUT is fetched successfully.
Auth-challenge is hashed using SHA-256 and transformed into base64-String.
Example: "OzvOBGndwenSMbCHB9Qe/rMrL1/bSsHOnMxKDy1wagI="

KOPS is called using Method "externalauthenticate" with signature-type "C.AUT", signature-schemes "RSASSA-PSS", card-handle of SMC-B and the base64 string. KOPS returns a token signed with the C-AUT-certificate from SMC-B.

GET Request with x-auth-signed-challenge (signed token) and x-auth-certificate (KOPS C-AUT) is sent. 
API returns an error page: **Ubirch Certify App Login, We are sorry... Certificate & Signature Verification failed.**

What triggers this error when the challenge was signed with the same certificate from the same SMC-B that is handed back to certify app?

**Comment by TH-Pega** [comment]
I'm partially successfull! With the special "Impfzentrum KoPS"-SMC-B and "client_id=user-access-ti-simulator" I get an access token now.

Unfortunately, the final step, calling https://api-pvs.certify.demo.ubirch.com/api/certify/v2/issue with "Authorization: Bearer <token>", "ContentType: application/vnd.dgc.v1+json" and the certificate data ends with error "HTTP 403 Forbidden" :-(

**Reply by alexanderwilms** [reply]
where do I get the Impfzentrum KoPS"-SMC-B?

**Reply by WePo64** [reply]
@alexanderwilms 
After signing the contract we got an email from Impfzertifikat-PVS (at) pd-g.de
This email contained the the two following attachments:
"Hinweise zum Impfnachweis für KoPS 3.1_V1.0.0.pdf"
"KoPS31_Testdaten_Impfnachweis_20210525.zip"
The second attachment contains the content of the SMC-B and the first attachment contains instructions how to integrate them into KoPS.

**Reply by norganim** [reply]
We got the same problem at the last calling [403] Forbidden (Invalid Credentials) Invalid JWT Token. 
We use the same parameter as @TH-Pega. When we test the JWT it's valid, so what is going wrong

**Comment by oliver-steinbrecher** [comment]
The client-id has been updated . Please use `user-access-ti-simulator` for testing in conjunction with Kops

**Reply by nicolaslang1** [reply]
Mh i still get the Error: 

We are sorry... 

Certificate & Signature Verification failed

On submitting the signed Challenge. In PU it works.

My OIDC Authorization Endpoint:
https://id.certify.demo.ubirch.com/auth/realms/ubirch-certify/protocol/openid-connect/auth

**Reply by Jan2Philip** [reply]
On our side we get a valid Certificate and Signature but the last step (Requesting the Certfificate) results in an HTTP 403 ( Invalid JWT Token)

As a card we use the provided smcb.

**Reply by oliver-steinbrecher** [reply]
we're still investigating ...

**Reply by norganim** [reply]
Hi are there any result of the investigation?

**Reply by Jan2Philip** [reply]
Withoud sending a bearer token we get a different Forbidden Message: "Sorry, the credential is invalid:ID not whitelisted"

**Reply by WePo64** [reply]
@oliver-steinbrecher 
Until the `api-pvs.certify.demo.ubirch.com` accepts the Bearer Token created with KoPS, we are **completely stuck in our progress**.
We can do nothing but wait for a solution.
Any news on the investigation?

**Comment by WePo64** [comment]
@oliver-steinbrecher 
The client-id user-access-ti-simulator works for us for steps 1-5 using KoPS.
But the final call to create the vaccination certificate against `api-pvs.certify.demo.ubirch.com` still results in:
`Error calling ApiCertifyV2IssuePost: Invalid JWT Token`
The PU works fine, but i think, we have to pass the Simulation first, before we can use the PU, or am i wrong here?
Can we skip the Simulation an get an approval to the contract for our solution just using the PU?

**Reply by oliver-steinbrecher** [reply]
There is an official admission process for PVS vendors - Tests can be done with the Kops or in the TI RU environment. The admission process also requires productive tests. Please get in touch with the official party to clarify this.

**Comment by oliver-steinbrecher** [comment]
We've updated the authentication endpoint - Please use this https://id.certify.demo.ubirch.com/auth/realms/bmg-ti-certify/protocol/openid-connect/auth . Flow has now been verified together with the gematik Kops .

**Reply by nicolaslang1** [reply]
I can confirm: This is working now. Thank you!

**Comment by oliver-steinbrecher** [comment]
Updated Endpoint for RU Testing environment details to be aligned with remaining documentation

**Comment by richSpiegel** [comment]
We still have the 401 Unauthorized error on both systems (with and without KOPS). Any hints?

Trigger endpoints are: 
https://id.certify.demo.ubirch.com/auth/realms/bmg-ti-certify/login-actions/authenticate 
and for KOPS
https://id.certify.demo.ubirch.com/auth/realms/bmg-ti-certify/login-actions/authenticate

Endpoint for receiving challenge without KOPS:   
https://id.certify.demo.ubirch.com/auth/realms/bmg-ti-certify/login-actions/authenticate

**Reply by oliver-steinbrecher** [reply]
@thinkberg please pick this up - sounds like the TSL is outdated

**Reply by thinkberg** [reply]
The TSL file up to date.

**Comment by richSpiegel** [comment]
Still not working (401 Unauthorized). Is our certificate and signature invalid?
The Response is:

<!DOCTYPE html
    PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" class="login-pf">

<head>
    <meta charset="utf-8">
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
    <meta name="robots" content="noindex, nofollow">

    <meta name="viewport" content="width=device-width,initial-scale=1" />
    <title>Log in to bmg-ti-certify</title>
    <link rel="icon" href="/auth/resources/mrpuj/login/Ubirch-bmg-kv/img/favicon.ico" />
    <link href="/auth/resources/mrpuj/common/keycloak/node_modules/patternfly/dist/css/patternfly.min.css"
        rel="stylesheet" />
    <link href="/auth/resources/mrpuj/common/keycloak/node_modules/patternfly/dist/css/patternfly-additions.min.css"
        rel="stylesheet" />
    <link href="/auth/resources/mrpuj/common/keycloak/lib/zocial/zocial.css" rel="stylesheet" />
    <link href="/auth/resources/mrpuj/login/Ubirch-bmg-kv/css/login.css" rel="stylesheet" />
    <SCRIPT>
        if (typeof history.replaceState === 'function') {  history.replaceState({}, "some title", "https://id.certify.demo.ubirch.com/auth/realms/bmg-ti-certify/login-actions/authenticate?execution=b5b9ee5f-6843-4074-9cba-4e52a79e991a&client_id=user-access-ti&tab_id=tMqEnM7illc"); }
    </SCRIPT>
</head>

<body class="">

    <div class="header">
        <img src="/auth/resources/mrpuj/login/Ubirch-bmg-kv/img/logo_kbv.png" alt="KV-Logo">
        <div id="kc-header-wrapper" class="">Anmeldemaske zur Zertifikatserstellung</div>
    </div>

    <div class="login-pf-page">
        <div id="kc-header" class="login-pf-page-header">
            <div id="kc-header-wrapper" class="">Anmeldemaske zur Zertifikatserstellung</div>
        </div>
        <div class="card-pf ">
            <header class="login-pf-header">
                <h1 id="kc-page-title"> We are sorry...
                </h1>
            </header>
            <div id="kc-content">
                <div id="kc-content-wrapper">

                    <div id="kc-error-message">
                        <p class="instruction">Certificate & Signature Verification failed</p>
                        <p><a id="backToApplication" href="https://certify.demo.ubirch.com">« Back to Application</a>
                        </p>
                    </div>

                </div>
            </div>

        </div>
    </div>
</body>

</html>

Regards

Richard
Von: Matthias L. Jugel ***@***.***
Gesendet: Freitag, 23. Juli 2021 09:56
An: Digitaler-Impfnachweis/certification-apis ***@***.***>
Cc: richSpiegel ***@***.***>; Comment ***@***.***>
Betreff: Re: [Digitaler-Impfnachweis/certification-apis] Testing the SMB-C authentication (#14)


The TSL file up to date.

**Comment by richSpiegel** [comment]
Hallo,

Ich habe gerade ein 504Gateway Timeout auf https://id.ru.impfnachweis.info/auth/realms/bmg-ti-certify/protocol/openid-connect/auth
Ist der Server down?

Mfg - Richard

Von: Matthias L. Jugel ***@***.***
Gesendet: Freitag, 23. Juli 2021 09:56
An: Digitaler-Impfnachweis/certification-apis ***@***.***>
Cc: richSpiegel ***@***.***>; Comment ***@***.***>
Betreff: Re: [Digitaler-Impfnachweis/certification-apis] Testing the SMB-C authentication (#14)


The TSL file up to date.

**Reply by jweko** [reply]
Hallo, 
bei uns ebenfalls das gleiche Probleme seit heute morgen 504 Gateway Timeout für die RU. 
VG
Julia

**Reply by richSpiegel** [reply]
I am using this endpoint now, which works:  https://id.certify.demo.ubirch.com

**Reply by richSpiegel** [reply]
still have an unauthorized 401 error though, when sending challenge...
